sound/soc/sof/virtio-be.c

```diff
@@ -5,122 +5,1039 @@
+{
+	struct snd_sof_ipc *ipc = sdev->ipc;
+	struct sof_ipc_hdr *hdr = (struct sof_ipc_hdr *)ipc_data;
+
+	return sof_ipc_tx_message(ipc, hdr->cmd, ipc_data, count,
+				  reply_data, reply_size);
+}
+
+/* Return true if posn buffer is filled successfully */
+static bool sbe_fill_posn_vqbuf(struct sof_vbe *vbe,
+				struct virtio_vq_info *vq,
+				struct sof_ipc_stream_posn *posn,
+				bool *endchain)
+{
+	struct device *dev = vbe->sdev->dev;
+	struct iovec iov;
+	u16 idx;
+	int ret;
+
+	*endchain = false;
+	while (virtio_vq_has_descs(vq)) {
+		ret = virtio_vq_getchain(vq, &idx, &iov, 1, NULL);
+		if (ret <= 0)
+			return false;
+
+		if (iov.iov_len < sizeof(struct sof_ipc_stream_posn)) {
+			*endchain = true;
+			dev_err(dev, "iov len %lu, expecting len %lu\n",
+				iov.iov_len, sizeof(*posn));
+			virtio_vq_relchain(vq, idx, iov.iov_len);
+			continue;
+		}
+		memcpy(iov.iov_base, posn, sizeof(struct sof_ipc_stream_posn));
+		*endchain = true;
+		virtio_vq_relchain(vq, idx, iov.iov_len);
+		return true;
+	}
+
+	return false;
+}
+
+/* IPC notification reply from FE to DSP */
+static void sbe_ipc_fe_not_reply_get(struct sof_vbe *vbe, int vq_idx)
+{
+	struct virtio_vq_info *vq;
+	struct vbs_sof_posn *entry;
+	unsigned long flags;
+	bool endchain, ret;
+
+	dev_dbg(vbe->sdev->dev,
+		"audio BE notification vq kick handling, vq_idx %d\n", vq_idx);
+
+	spin_lock_irqsave(&vbe->posn_lock, flags);
+	if (list_empty(&vbe->posn_list)) {
+		/* nothing to do */
+		spin_unlock_irqrestore(&vbe->posn_lock, flags);
+		return;
+	}
+	vq = &vbe->vqs[vq_idx];
+	entry = list_first_entry(&vbe->posn_list,
+				 struct vbs_sof_posn, list);
+
+	ret = sbe_fill_posn_vqbuf(vbe, vq, &entry->pos, &endchain);
+	spin_unlock_irqrestore(&vbe->posn_lock, flags);
+	if (endchain)
+		virtio_vq_endchains(vq, 1);
+}
+
+/* validate component IPC */
+static int sbe_ipc_comp(struct snd_sof_dev *sdev, int vm_id,
+			struct sof_ipc_hdr *hdr)
+{
+	/*TODO validate host comp id range based on vm_id */
+
+	/* Nothing to be done */
+	return 0;
+}
+
+/*
+ * This function is to get the BE dai link for the GOS
+ * It uses the dai_link name to find the BE dai link.
+ * The Current dai_link name "vm_dai_link" is for the GOS,
+ * which means only one Virtual Machine is supported.
+ * And the VM only support one playback pcm and one capture pcm.
+ * After we switch to the new topology, we can support multiple
+ * VMs and multiple PCM streams for each VM.
+ * This function may be abandoned after switching to the new topology.
  */
+static struct snd_pcm_substream *
+sbe_get_substream(struct snd_sof_dev *sdev,
+		  struct snd_soc_pcm_runtime **rtd, int direction)
+{
+	struct snd_pcm *pcm;
+	struct snd_pcm_str *stream;
+	struct snd_pcm_substream *substream = NULL;
+	struct snd_soc_dai_link *dai_link;
+	struct snd_soc_card *card = sdev->card;
+	struct snd_soc_pcm_runtime *r;
+
+	list_for_each_entry(r, &card->rtd_list, list) {
+		pcm = r->pcm;
+		if (!pcm)
+			continue;
+
+		stream = &pcm->streams[direction];
```

Comment:
again, more comments - i know there is a big comment at the top, but I need to see the conditions you are checking at every change of code flow.  so above this statement it would be /* check stream direction matches */